addons/account/models/account_reconcile_model.py

```diff
@@ -802,26 +802,41 @@ def _get_invoice_matching_rule_result(self, st_line, candidates, aml_ids_to_excl
         new_treated_aml_ids = set()
         candidates, priorities = self._filter_candidates(candidates, aml_ids_to_exclude, reconciled_amls_ids)
 
-        # Special case: the amounts are the same, submit the line directly.
         st_line_currency = st_line.foreign_currency_id or st_line.currency_id
-        candidate_currencies = set(candidate['aml_currency_id'] or st_line.company_id.currency_id.id for candidate in candidates)
+        candidate_currencies = set(candidate['aml_currency_id'] for candidate in candidates)
+        kept_candidates = candidates
         if candidate_currencies == {st_line_currency.id}:
+            kept_candidates = []
+            sum_kept_candidates = 0
             for candidate in candidates:
-                residual_amount = candidate['aml_currency_id'] and candidate['aml_amount_residual_currency'] or candidate['aml_amount_residual']
-                if st_line_currency.is_zero(residual_amount + st_line.amount_residual):
-                    candidates, priorities = self._filter_candidates([candidate], aml_ids_to_exclude, reconciled_amls_ids)
+                candidate_residual = candidate['aml_amount_residual_currency']
+
+                if st_line_currency.is_zero(candidate_residual + st_line.amount_residual):
```

Comment:
This part is wrong as the previous behavior in case of `same_currency` is disabled :/

Reply:
https://pbs.twimg.com/media/DWKDxlYW0AE2BMG.jpg

Reply:
```suggestion
                if st_line_currency.compare_amounts(sum_kept_candidates, st_line.amount_residual) == 0:
```

Because https://github.com/odoo/odoo/blob/dfce5361a03214f456727d68dff136c81d84f37a/odoo/addons/base/models/res_currency.py#L203-L207

I also don't get why this line is comparing raw values, and the line a bit below is comparing absolute values. If it is intended, it deserves a comment.
If it is intended, I think it will cause an issue with https://github.com/odoo/odoo/blob/cb0ab1e91ea515615d8bab3f62f5896400875cec/addons/account/models/account_reconcile_model.py#L824 because that line is not considering the absolute values.

Reply:
Absolue values are actually used here to avoid making different cases for positive or negative balance on the statement line. Since we know for sure (thanks to this https://github.com/odoo/odoo/blame/saas-14.5/addons/account/models/account_reconcile_model.py#L622) that the statement lines and candidates have the same balance sign, this works fine. I'll see what I can do for the comment.